src/main/scala/sigmastate/eval/Evaluation.scala

```diff
@@ -84,58 +89,161 @@ trait Evaluation extends RuntimeCosting { IR: IRContext =>
+
+  def isAllowedOpCodeInCosting(opCode: OpCode): Boolean =
+    (opCode == Undefined) || allowedOpCodesInCosting.contains(opCode)
+
+  def isNonSerializable(d: Def[_]): Boolean = d match {
     case _: OpCost | _: PerKbCostOf | _: Cast[_] => true
-    case _: Tup[_,_] | _: First[_,_] | _: Second[_,_] => true
-    case _: FieldApply[_] => true
     case _: IntPlusMonoid => true
-    case _: Lambda[_,_] => true
     case _: ThunkDef[_] => true
-    case ApplyUnOp(_: NumericToLong[_] | _: NumericToInt[_], _) => true
-    case ApplyBinOp(_: NumericPlus[_] | _: NumericTimes[_] | _: OrderingMax[_] | _: IntegralDivide[_] ,_,_) => true
-
-    case SCM.inputs(_) | SCM.outputs(_) | SCM.dataInputs(_) | SCM.selfBox(_) | SCM.lastBlockUtxoRootHash(_) | SCM.headers(_) |
-         SCM.preHeader(_) | SCM.getVar(_,_,_) => true
-    case SBM.propositionBytes(_) |  SBM.bytes(_) |  SBM.bytesWithoutRef(_) |  SBM.registers(_) |  SBM.getReg(_,_,_) |
-         SBM.tokens(_)  => true
+    case SCM.inputs(_) | SCM.outputs(_) | SCM.dataInputs(_) | SCM.selfBox(_) | SCM.lastBlockUtxoRootHash(_) |
+         SCM.headers(_) | SCM.preHeader(_) | SCM.getVar(_, _, _) => true
+    case SBM.propositionBytes(_) | SBM.bytes(_) | SBM.bytesWithoutRef(_) | SBM.registers(_) |
+         SBM.getReg(_, _, _) | SBM.tokens(_) => true
     case SSPM.propBytes(_) => true
     case SAVM.tVal(_) | SAVM.valueSize(_) => true
     case SizeM.dataSize(_) => true
     case SPairM.l(_) | SPairM.r(_) => true
     case SCollM.sizes(_) => true
     case SOptM.sizeOpt(_) => true
     case SFuncM.sizeEnv(_) => true
-    case _: CSizePairCtor[_,_] | _: CSizeFuncCtor[_,_,_] | _: CSizeOptionCtor[_] | _: CSizeCollCtor[_] |
+    case _: CSizePairCtor[_, _] | _: CSizeFuncCtor[_, _, _] | _: CSizeOptionCtor[_] | _: CSizeCollCtor[_] |
          _: CSizeBoxCtor | _: CSizeContextCtor | _: CSizeAnyValueCtor => true
-    case ContextM.SELF(_) | ContextM.OUTPUTS(_) | ContextM.INPUTS(_) | ContextM.dataInputs(_) | ContextM.LastBlockUtxoRootHash(_) |
-         ContextM.getVar(_,_,_)  => true
-    case SigmaM.propBytes(_) => true
-    case _: CReplCollCtor[_] | _: PairOfColsCtor[_,_] => true
-    case CollM.length(_) | CollM.map(_,_) | CollM.sum(_,_) | CollM.zip(_,_) | CollM.slice(_,_,_) | CollM.apply(_,_) |
-         CollM.append(_,_) | CollM.foldLeft(_,_,_) => true
-    case CBM.replicate(_,_,_) | CBM.fromItems(_,_,_) => true
-    case BoxM.propositionBytes(_) | BoxM.bytesWithoutRef(_) | BoxM.getReg(_,_,_) => true
-    case OM.get(_) | OM.getOrElse(_,_) | OM.fold(_,_,_) | OM.isDefined(_) => true
+    case _: CReplCollCtor[_] | _: PairOfColsCtor[_, _] => true
+    case CollM.sum(_, _) => true
+    case CBM.replicate(_, _, _) | CBM.fromItems(_, _, _) => true
     case _: CostOf | _: SizeOf[_] => true
-    case _: Upcast[_,_] => true
-    case _: Apply[_,_] => true
     case SPCM.some(_) => true
     case _ => false
   }
 
+  def getOpCode(d: Def[_]): OpCode = if (isNonSerializable(d)) Undefined else d match {
```

Comment:
isNonSerializable check should be new ValidationRule which extends SoftForkWhenCodeAdded.

Reply:
@aslesarenko Damn it, I had the same thoughts back then but dismissed them as being too complicated. :)